File: backend/src/services/role.service.ts
import { roleProvider } from '../providers/roleProvider.js';
import { RoleRes } from '../schemas/role.js';
import { BaseListResType, BaseSearchRequestType } from '../schemas/search.js';
import { AppError } from '../utils/AppError.js';
import type {
  RoleCreateBodyType,
  RoleUpdateBodyType,
  RoleResType,
} from '../schemas/role.js';

export const roleService = {
  async getAll(request: BaseSearchRequestType): Promise<BaseListResType> {
    const limit = request.limit;
    const rows = await roleProvider.findAll(request);
    const hasNextPage = rows.length > limit;
    const items = hasNextPage ? rows.slice(0, limit) : rows;
    const nextCursor =
      hasNextPage && items.length > 0 ? items[items.length - 1].id : null;

    return {
      data: items.map((row) => RoleRes.parse(row)),
      limit,
      nextCursor,
      hasNextPage,
    };
  },

  async findById(id: number) {
    const row = await roleProvider.findById(id);
    if (!row) throw new AppError(404, 'Role not found');
    return RoleRes.parse(row);
  },

  async create(data: RoleCreateBodyType): Promise<RoleResType> {
    const existing = await roleProvider.findRoleByName(data.name);
    if (existing) throw new AppError(409, 'Role already exists');
    const row = await roleProvider.create({
      name: data.name,
      description: data.description ?? null,
    });
    return RoleRes.parse(row);
  },

  async update(data: RoleUpdateBodyType, id: number): Promise<RoleResType> {
    const existing = await roleProvider.findById(id);
    if (!existing) throw new AppError(404, 'Role not found');
    const row = await roleProvider.update(
      {
        ...(data.name !== undefined ? { name: data.name } : {}),
        ...(data.description !== undefined
          ? { description: data.description }
          : {}),
        ...(data.active !== undefined ? { active: data.active } : {}),
      },
      id,
    );
    return RoleRes.parse(row);
  },

  async delete(id: number): Promise<RoleResType> {
    const existing = await roleProvider.findById(id);
    if (!existing) throw new AppError(404, 'Role not found');
    const row = await roleProvider.delete(id);
    return RoleRes.parse(row);
  },
};

File: backend/app/services/role_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from app.providers.role_provider import role_provider
from app.utils.app_error import AppError
from typing import Optional

async def get_all(db: AsyncSession, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None) -> dict:
    rows = await role_provider.find_all(db, limit=limit, cursor=cursor, order=order, search=search)
    has_next = len(rows) > limit
    items = rows[:limit] if has_next else rows
    next_cursor = items[-1].id if has_next and items else None
    return {"data": items, "limit": limit, "nextCursor": next_cursor, "hasNextPage": has_next}

async def find_by_id(db: AsyncSession, role_id: int):
    role = await role_provider.find_by_id(db, role_id)
    if not role:
        raise AppError(404, "Role not found")
    return role

async def create(db: AsyncSession, name: str, description: Optional[str] = None):
    return await role_provider.create(db, name=name, description=description)

async def update(db: AsyncSession, role_id: int, **data):
    role = await role_provider.update(db, role_id, **{k: v for k, v in data.items() if v is not None})
    if not role:
        raise AppError(404, "Role not found")
    return role

async def delete(db: AsyncSession, role_id: int):
    role = await role_provider.soft_delete(db, role_id)
    if not role:
        raise AppError(404, "Role not found")
    return role
